include/aos/common/tools/memory.hpp

```diff
@@ -122,66 +175,136 @@ class SmartPtr {
  *
  * @tparam T holding object type.
  */
-template <typename T>
-class UniquePtr : public SmartPtr<T>, private NonCopyable {
+template <typename T, typename Deleter = DefaultDeleter<T>>
```

Comment:
shared_ptr also may have custom deleter, so both smart pointers should have similar functionality

Reply:
There is a subtle difference between deleter in  unique & shared pointers, that makes it difficult to have a common implementation. Deleter for unique_ptr is stored on the stack, while shared_ptr's deleter stored in the heap(allocator's space in our case). SmartPtr implements a shared_ptr's approach, so I made a separate impl for unique_ptr.
 
There is also a difference in behaviour for the following example:
struct A{}; struct B: A {}
unique_ptr<A> b = make_unique<B>();
shared_ptr<A> b = make_shared<B>();

shared_ptr calls B & A destructors while unique_ptr - only A's destructor. 

There should also be a difference in interface between shared_ptr & unique_ptr: in our case shared_ptr should take a reference to allocator to allocate service part(weak/shared counters, optional mutex & deleter). unique_ptr doesn't require that.

Reply:
> There is a subtle difference between deleter in unique & shared pointers, that makes it difficult to have a common implementation. Deleter for unique_ptr is stored on the stack, while shared_ptr's deleter stored in the heap(allocator's space in our case). SmartPtr implements a shared_ptr's approach, so I made a separate impl for unique_ptr.
> 
> There is also a difference in behaviour for the following example:
> 
> ```
> struct A{}; struct B: A {}
> unique_ptr<A> b = make_unique<B>();
> shared_ptr<A> b = make_shared<B>();
> ```
> 
> shared_ptr calls B & A destructors while unique_ptr - only A's destructor.
> 
> There should also be a difference in interface between shared_ptr & unique_ptr: in our case shared_ptr should take a reference to allocator to allocate service part(weak/shared counters, optional mutex & deleter). unique_ptr doesn't require that.

I did not get, in which compiler do this while unique_ptr - only A's destructor.? according to https://godbolt.org/, this call works correct  https://godbolt.org/z/qGvEn83E9
```
Base
Derived
```

Reply:
> There is a subtle difference between deleter in unique & shared pointers, that makes it difficult to have a common implementation. Deleter for unique_ptr is stored on the stack, while shared_ptr's deleter stored in the heap(allocator's space in our case). SmartPtr implements a shared_ptr's approach, so I made a separate impl for unique_ptr.
> 
> There is also a difference in behaviour for the following example:
> 
> ```
> struct A{}; struct B: A {}
> unique_ptr<A> b = make_unique<B>();
> shared_ptr<A> b = make_shared<B>();
> ```
> 
> shared_ptr calls B & A destructors while unique_ptr - only A's destructor.
> 
> There should also be a difference in interface between shared_ptr & unique_ptr: in our case shared_ptr should take a reference to allocator to allocate service part(weak/shared counters, optional mutex & deleter). unique_ptr doesn't require that.

@mykola-kobets-epam  I did not get, why you decide unique_ptr - only A's destructor? https://godbolt.org/z/hrjev1v48 works well 
```
#include <iostream>
#include <string>
#include <memory>

struct Base {
   Base() {
      std::cout << "Base\n";
   }
};

struct Derived: Base {
   Derived() {
      std::cout << "Derived\n";
   }
};


int main() {
    std::unique_ptr<Base> ptr = std::make_unique<Derived>();
    return 0;
}

Base
Derived
```

Reply:
SharedPtr refactoring should be done in a separate PR with possible implementation of weakptr